[quant] Attach qconfig to all modules

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42576 [quant] Attach qconfig to all modules**

Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

jerryzh168 committed:
[quant] Attach qconfig to all modules

Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on torch/quantization/quantize.py:
```diff
@@ -42,8 +42,7 @@ def _propagate_qconfig_helper(module, qconfig_dict, white_list=None,
 
-    if type(module) in white_list:
-        module.qconfig = module_qconfig
+    module.qconfig = module_qconfig
```
How do we make sure only the ones that are allowed are getting the qconfig?

jerryzh168 replied:
yeah, I realized that, just updated the diff. Ideally we don't have people using this "white_list" API, otherwise we'll need to make changes to callsite.

raghuramank100 replied:
The numeric suite tests might be impacted by this change, please check if they are passing prior to landing.

raghuramank100 replied:
Also, what is the motivation of this change? Why do we need a qconfig for all modules, even if they are not observed?

vkuzo commented:
btw, what's the context for this change?  Is this for GraphModule, for adding finer grain control such as "don't quantize stateless aten op X", something else?

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]

jerryzh168 commented:
> btw, what's the context for this change? Is this for GraphModule, for adding finer grain control such as "don't quantize stateless aten op X", something else?

Yeah this is for graphmodule, we need to attach qconfig to the parent_module to be able to quantize the ops in the forward of that module. for example:

```python
class Linear(torch.nn.Module):
	def __init__(self, weight):
		super().__init__()

	def forward(self, x, weight):
		return F.linear(x, weight)
```

vkuzo commented on torch/quantization/quantize.py:
```diff
@@ -130,7 +129,8 @@ def add_observer_(module, non_leaf_module_list=None, device=None, prehook=None):
     if hasattr(module, 'qconfig') and module.qconfig is not None and \
-       len(module._modules) == 0 and not isinstance(module, torch.nn.Sequential):
+       len(module._modules) == 0 and not isinstance(module, torch.nn.Sequential) \
+       and type(module) in DEFAULT_QCONFIG_PROPAGATe_WHITE_LIST:
```
nit: `PROPAGATE` instead of `PROPAGATe`?

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -130,7 +129,8 @@ def add_observer_(module, non_leaf_module_list=None, device=None, prehook=None):
     if hasattr(module, 'qconfig') and module.qconfig is not None and \
-       len(module._modules) == 0 and not isinstance(module, torch.nn.Sequential):
+       len(module._modules) == 0 and not isinstance(module, torch.nn.Sequential) \
+       and type(module) in DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST:
```
We should be using white_list instead of DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST

raghuramank100 reviewed (changes requested):
Had a few questions. Thanks!

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Attach qconfig to all modules"


Summary:
Previously we have qconfig propagate list and we only attach qconfig for modules
in the list, this works when everything is quantized in the form of module.
but now we are expanding quantization for functional/torch ops, we'll need to attach qconfig
to all modules

Test Plan:
regression tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22939453](https://our.internmc.facebook.com/intern/diff/D22939453)

[ghstack-poisoned]
